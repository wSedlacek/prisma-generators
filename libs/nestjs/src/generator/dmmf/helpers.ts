// TODO: support not string args
export const modelAttributeRegex = /(@@NestJS\.)+([A-z])+(\(")+([A-z])+("\))+/;
export const fieldAttributeRegex = /(@NestJS\.)+([A-z])+(\(")+([A-z])+("\))+/;
export const attributeNameRegex = /(?:\.)+([A-Za-z])+(?:\()+/;
export const attributeArgsRegex = /(?:\(")+([A-Za-z])+(?:"\))+/;

export const parseDocumentationAttributes = (
  documentation: string | undefined,
  expectedAttributeName: string,
  expectedAttributeKind: 'model' | 'field'
) => {
  const attributeRegex =
    expectedAttributeKind === 'model'
      ? modelAttributeRegex
      : fieldAttributeRegex;
  const attribute = documentation?.match(attributeRegex)?.[0];
  const attributeName = attribute?.match(attributeNameRegex)?.[0]?.slice(1, -1);
  if (attributeName !== expectedAttributeName) {
    return undefined;
  }

  return attribute?.match(attributeArgsRegex)?.[0]?.slice(1, -1);
};
