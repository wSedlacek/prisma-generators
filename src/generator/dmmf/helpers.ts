// TODO: support not string args
export const modelAttributeRegex = /(@@TypeGraphQL\.)+([A-z])+(\(")+([A-z])+("\))+/;
export const fieldAttributeRegex = /(@TypeGraphQL\.)+([A-z])+(\(")+([A-z])+("\))+/;
export const attributeNameRegex = /(?:\.)+([A-Za-z])+(?:\()+/;
export const attributeArgsRegex = /(?:\(")+([A-Za-z])+(?:"\))+/;

export function parseDocumentationAttributes(
  documentation: string | undefined,
  expectedAttributeName: string,
  expectedAttributeKind: "model" | "field",
) {
  const attributeRegex =
    expectedAttributeKind === "model"
      ? modelAttributeRegex
      : fieldAttributeRegex;
  const attribute = documentation?.match(attributeRegex)?.[0];
  const attributeName = attribute?.match(attributeNameRegex)?.[0]?.slice(1, -1);
  if (attributeName !== expectedAttributeName) {
    return;
  }
  const attributeArgs = attribute?.match(attributeArgsRegex)?.[0]?.slice(1, -1);
  return attributeArgs;
}
