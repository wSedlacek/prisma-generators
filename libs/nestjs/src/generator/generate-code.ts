import { DMMF as PrismaDMMF } from '@prisma/client/runtime/dmmf-types';
import path from 'path';
import { Project } from 'ts-morph';

import { isDefined, saveSourceFile } from '../utils';
import { generateArgsTypeClassFromArgs } from './args-class';
import {
  argsFolderName,
  crudResolversFolderName,
  enumsFolderName,
  inputsFolderName,
  modelsFolderName,
  outputsFolderName,
  relationsResolversFolderName,
  resolversFolderName,
} from './config';
import { DmmfDocument } from './dmmf/dmmf-document';
import { DMMF } from './dmmf/types';
import { generateEnumFromDef } from './enum';
import { noop } from './helpers';
import {
  generateArgsBarrelFile,
  generateEnumsBarrelFile,
  generateIndexFile,
  generateInputsBarrelFile,
  generateModelsBarrelFile,
  generateOutputsBarrelFile,
  generateResolversBarrelFile,
} from './imports';
import { generateObjectTypeClassFromModel } from './model-type-class';
import { GenerateCodeOptions } from './options';
import { generateCrudResolverClassFromMapping } from './resolvers/full-crud';
import { generateRelationsResolverClassesFromModel } from './resolvers/relations';
import {
  generateInputTypeClassFromType,
  generateOutputTypeClassFromType,
} from './type-class';

export const generateCode = async (
  dmmf: PrismaDMMF.Document,
  options: GenerateCodeOptions,
  log: (msg: string) => void = noop
) => {
  const baseDirPath = options.outputDirPath;
  const project = new Project();
  const resolversDirPath = path.resolve(baseDirPath, resolversFolderName);
  const modelNames = dmmf.datamodel.models.map((model) => model.name);

  log('Transforming dmmfDocument...');
  const dmmfDocument = new DmmfDocument(dmmf, options);

  log('Generating enums...');
  const datamodelEnumNames = dmmfDocument.datamodel.enums.map(
    (enumDef) => enumDef.typeName
  );
  await Promise.all(
    dmmfDocument.datamodel.enums.map((enumDef) =>
      generateEnumFromDef(project, baseDirPath, enumDef)
    )
  );
  await Promise.all(
    dmmfDocument.schema.enums
      // skip enums from datamodel
      .filter((enumDef) => !datamodelEnumNames.includes(enumDef.typeName))
      .map((enumDef) => generateEnumFromDef(project, baseDirPath, enumDef))
  );
  const emittedEnumNames = [
    ...new Set([
      ...dmmfDocument.schema.enums.map((it) => it.typeName),
      ...dmmfDocument.datamodel.enums.map((it) => it.typeName),
    ]),
  ];
  const enumsBarrelExportSourceFile = project.createSourceFile(
    path.resolve(baseDirPath, enumsFolderName, 'index.ts'),
    undefined,
    { overwrite: true }
  );
  generateEnumsBarrelFile(enumsBarrelExportSourceFile, emittedEnumNames);
  await saveSourceFile(enumsBarrelExportSourceFile);

  log('Generating models...');
  await Promise.all(
    dmmfDocument.datamodel.models.map((model) =>
      generateObjectTypeClassFromModel(
        project,
        baseDirPath,
        model,
        dmmfDocument,
        options
      )
    )
  );
  const modelsBarrelExportSourceFile = project.createSourceFile(
    path.resolve(baseDirPath, modelsFolderName, 'index.ts'),
    undefined,
    { overwrite: true }
  );
  generateModelsBarrelFile(
    modelsBarrelExportSourceFile,
    dmmfDocument.datamodel.models.map((it) => it.typeName)
  );
  await saveSourceFile(modelsBarrelExportSourceFile);

  log('Generating output types...');
  const rootTypes = dmmfDocument.schema.outputTypes.filter((type) =>
    ['Query', 'Mutation'].includes(type.name)
  );
  const outputTypesToGenerate = dmmfDocument.schema.outputTypes.filter(
    // skip generating models and root resolvers
    (type) => !modelNames.includes(type.name) && !rootTypes.includes(type)
  );
  const outputTypesFieldsArgsToGenerate = outputTypesToGenerate
    .map((it) => it.fields)
    .reduce((a, b) => a.concat(b), [])
    .filter((it): it is Required<DMMF.OutputSchemaField> => !!it.argsTypeName);
  await Promise.all(
    outputTypesToGenerate.map((type) =>
      generateOutputTypeClassFromType(
        project,
        resolversDirPath,
        type,
        dmmfDocument,
        options
      )
    )
  );
  const outputsBarrelExportSourceFile = project.createSourceFile(
    path.resolve(
      baseDirPath,
      resolversFolderName,
      outputsFolderName,
      'index.ts'
    ),
    undefined,
    { overwrite: true }
  );
  generateOutputsBarrelFile(
    outputsBarrelExportSourceFile,
    outputTypesToGenerate.map((it) => it.typeName),
    outputTypesFieldsArgsToGenerate.length > 0
  );
  await saveSourceFile(outputsBarrelExportSourceFile);

  if (outputTypesFieldsArgsToGenerate.length > 0) {
    log('Generating output types args...');
    await Promise.all(
      outputTypesFieldsArgsToGenerate.map(async (field) => {
        await generateArgsTypeClassFromArgs(
          project,
          path.resolve(resolversDirPath, outputsFolderName),
          field.args,
          field.argsTypeName,
          dmmfDocument,
          options,
          2
        );
      })
    );
    const outputsArgsBarrelExportSourceFile = project.createSourceFile(
      path.resolve(
        baseDirPath,
        resolversFolderName,
        outputsFolderName,
        argsFolderName,
        'index.ts'
      ),
      undefined,
      { overwrite: true }
    );
    generateArgsBarrelFile(
      outputsArgsBarrelExportSourceFile,
      outputTypesFieldsArgsToGenerate
        .map((it) => it.argsTypeName)
        .filter(isDefined)
    );
    await saveSourceFile(outputsArgsBarrelExportSourceFile);
  }

  log('Generating input types...');
  await Promise.all(
    dmmfDocument.schema.inputTypes.map((type) =>
      generateInputTypeClassFromType(
        project,
        resolversDirPath,
        type,
        dmmfDocument,
        options
      )
    )
  );
  const inputsBarrelExportSourceFile = project.createSourceFile(
    path.resolve(
      baseDirPath,
      resolversFolderName,
      inputsFolderName,
      'index.ts'
    ),
    undefined,
    { overwrite: true }
  );
  generateInputsBarrelFile(
    inputsBarrelExportSourceFile,
    dmmfDocument.schema.inputTypes.map((it) => it.typeName)
  );
  await saveSourceFile(inputsBarrelExportSourceFile);

  log('Generating relation resolvers...');
  const relationResolversData = await Promise.all(
    dmmfDocument.datamodel.models
      .filter((model) => model.fields.some((field) => field.relationName))
      .map(async (model) => {
        const outputType = dmmfDocument.schema.outputTypes.find(
          (type) => type.name === model.name
        );
        const mapping = dmmfDocument.mappings.find(
          (it) => it.model === model.name
        );

        if (!outputType || !mapping) {
          throw new Error('Could not determine resolver data');
        }

        return generateRelationsResolverClassesFromModel(
          project,
          baseDirPath,
          model,
          mapping,
          outputType,
          dmmfDocument,
          options
        );
      })
  );
  if (relationResolversData.length > 0) {
    const relationResolversBarrelExportSourceFile = project.createSourceFile(
      path.resolve(
        baseDirPath,
        resolversFolderName,
        relationsResolversFolderName,
        'index.ts'
      ),
      undefined,
      { overwrite: true }
    );
    generateResolversBarrelFile(
      'relations',
      relationResolversBarrelExportSourceFile,
      relationResolversData
    );
    await saveSourceFile(relationResolversBarrelExportSourceFile);
  }

  log('Generating crud resolvers...');
  const crudResolversData = await Promise.all(
    dmmfDocument.mappings.map(async (mapping) => {
      const model = dmmfDocument.datamodel.models.find(
        (it) => it.name === mapping.model
      );

      if (!model) {
        throw new Error('Could not determine resolver data');
      }

      return generateCrudResolverClassFromMapping(
        project,
        baseDirPath,
        mapping,
        model,
        modelNames,
        options,
        dmmfDocument
      );
    })
  );
  const crudResolversBarrelExportSourceFile = project.createSourceFile(
    path.resolve(
      baseDirPath,
      resolversFolderName,
      crudResolversFolderName,
      'index.ts'
    ),
    undefined,
    { overwrite: true }
  );
  generateResolversBarrelFile(
    'crud',
    crudResolversBarrelExportSourceFile,
    crudResolversData
  );
  await saveSourceFile(crudResolversBarrelExportSourceFile);

  log('Generating index file');
  const indexSourceFile = project.createSourceFile(
    `${baseDirPath}/index.ts`,
    undefined,
    { overwrite: true }
  );
  generateIndexFile(indexSourceFile, relationResolversData.length > 0);
  await saveSourceFile(indexSourceFile);
};
