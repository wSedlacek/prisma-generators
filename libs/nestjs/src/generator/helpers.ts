import { DmmfDocument } from './dmmf/dmmf-document';
import { fieldAttributeRegex, modelAttributeRegex } from './dmmf/helpers';
import { DMMF } from './dmmf/types';

export const noop = () => {};

export const getFieldTSType = (
  typeInfo: DMMF.TypeInfo,
  dmmfDocument: DmmfDocument,
  isInputType: boolean,
  modelName?: string,
  typeName?: string
) => {
  let TSType: string;
  if (typeInfo.kind === 'scalar') {
    TSType = mapScalarToTSType(typeInfo.type, isInputType);
  } else if (typeInfo.kind === 'object') {
    TSType =
      dmmfDocument.getModelTypeName(typeInfo.type) ?? (!typeName || !modelName)
        ? getInputTypeName(typeInfo.type, dmmfDocument)
        : typeInfo.type.replace(modelName, typeName);
  } else if (typeInfo.kind === 'enum') {
    TSType = `keyof typeof ${typeInfo.type}`;
  } else {
    throw new Error(`Unsupported field type kind: ${typeInfo.kind}`);
  }
  if (typeInfo.isList) {
    if (TSType.includes(' ')) {
      TSType = `Array<${TSType}>`;
    } else {
      TSType += '[]';
    }
  }
  if (!typeInfo.isRequired) {
    // FIXME: use properly null for output and undefined for input
    // TSType += ' | undefined';
    TSType += ' | null | undefined';
  }

  return TSType;
};

export const mapScalarToTSType = (scalar: string, isInputType: boolean) => {
  switch (scalar) {
    case 'ID':
    case 'UUID': {
      return 'string';
    }
    case 'String': {
      return 'string';
    }
    case 'Boolean': {
      return 'boolean';
    }
    case 'DateTime': {
      return 'Date';
    }
    case 'Int':
    case 'Float': {
      return 'number';
    }
    case 'Json':
      return isInputType ? 'InputJsonValue' : 'JsonValue';
    default:
      throw new Error(`Unrecognized scalar type: ${scalar}`);
  }
};

export const getGraphQLType = (
  typeInfo: DMMF.TypeInfo,
  dmmfDocument: DmmfDocument,
  modelName?: string,
  typeName?: string
) => {
  let GraphQLType: string;
  if (typeInfo.kind === 'scalar') {
    GraphQLType = mapScalarToType(typeInfo.type);
  } else if (typeInfo.kind === 'object') {
    GraphQLType =
      dmmfDocument.getModelTypeName(typeInfo.type) ?? (!typeName || !modelName)
        ? getInputTypeName(typeInfo.type, dmmfDocument)
        : typeInfo.type.replace(modelName, typeName);
  } else {
    GraphQLType = typeInfo.type;
  }
  if (typeInfo.isList) {
    GraphQLType = `[${GraphQLType}]`;
  }

  return GraphQLType;
};

export const mapScalarToType = (scalar: string) => {
  switch (scalar) {
    case 'DateTime': {
      return 'Date';
    }
    // TODO: use proper uuid graphql scalar
    case 'UUID': {
      return 'String';
    }
    case 'Boolean':
    case 'String': {
      return scalar;
    }
    case 'ID':
    case 'Int':
    case 'Float': {
      return `${scalar}`;
    }
    case 'Json': {
      return `GraphQLJSON`;
    }
    default: {
      throw new Error(`Unrecognized scalar type: ${scalar}`);
    }
  }
};

export const camelCase = (str: string) => {
  return str[0].toLowerCase() + str.slice(1);
};

export const pascalCase = (str: string): string => {
  return str[0].toUpperCase() + str.slice(1);
};

const getInputKeywordPhrasePosition = (inputTypeName: string) => {
  const inputParseResult = [
    'Create',
    'OrderBy',
    'Update',
    'Upsert',
    'ScalarWhere',
    'Where',
    'ListRelationFilter',
    'RelationFilter',
    'Filter',
  ]
    .map((inputKeyword) => inputTypeName.search(inputKeyword))
    .filter((position) => position >= 0);

  if (inputParseResult.length === 0) {
    return undefined;
  }

  return inputParseResult[0];
};

export const getModelNameFromInputType = (inputTypeName: string) => {
  const keywordPhrasePosition = getInputKeywordPhrasePosition(inputTypeName);
  if (!keywordPhrasePosition) {
    return undefined;
  }

  return inputTypeName.slice(0, keywordPhrasePosition);
};

export const getInputTypeName = (
  originalInputName: string,
  dmmfDocument: DmmfDocument
): string => {
  const keywordPhrasePosition = getInputKeywordPhrasePosition(
    originalInputName
  );
  if (!keywordPhrasePosition) {
    return originalInputName;
  }

  const modelName = originalInputName.slice(0, keywordPhrasePosition);
  const typeNameRest = originalInputName.slice(keywordPhrasePosition);
  const modelTypeName = dmmfDocument.getModelTypeName(modelName);
  if (!modelTypeName) {
    return originalInputName;
  }

  return `${modelTypeName}${typeNameRest}`;
};

export const cleanDocsString = (
  documentation: string | undefined
): string | undefined => {
  if (!documentation) {
    return undefined;
  }
  let cleanedDocs = documentation;
  cleanedDocs = cleanedDocs.replace(modelAttributeRegex, '');
  cleanedDocs = cleanedDocs.replace(fieldAttributeRegex, '');
  cleanedDocs = cleanedDocs.split('"').join('\\"');
  cleanedDocs = cleanedDocs.split('\r').join('');
  cleanedDocs = cleanedDocs.split('\\r').join('');
  cleanedDocs = cleanedDocs.split('\n').join('');
  cleanedDocs = cleanedDocs.split('\\n').join('');

  return cleanedDocs;
};

export const toUnixPath = (maybeWindowsPath: string) => {
  return maybeWindowsPath.split('\\').join('/');
};
