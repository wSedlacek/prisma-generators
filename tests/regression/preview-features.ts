import { promises as fs } from "fs";

import generateArtifactsDirPath from "../helpers/artifacts-dir";
import { generateCodeFromSchema } from "../helpers/generate-code";
import createReadGeneratedFile, {
  ReadGeneratedFile,
} from "../helpers/read-file";

describe("preview features", () => {
  let outputDirPath: string;
  let readGeneratedFile: ReadGeneratedFile;

  beforeEach(async () => {
    outputDirPath = generateArtifactsDirPath("preview-features");
    await fs.mkdir(outputDirPath, { recursive: true });
    readGeneratedFile = createReadGeneratedFile(outputDirPath);
  });

  describe("when `distinct` is enabled", () => {
    it("should properly generate args classes and enums for model actions resolvers", async () => {
      const schema = /* prisma */ `
        model User {
          intIdField          Int      @id @default(autoincrement())
          uniqueStringField   String   @unique
          optionalFloatField  Float?
          dateField           DateTime
        }
      `;

      await generateCodeFromSchema(schema, {
        outputDirPath,
        enabledPreviewFeatures: ["distinct"],
      });
      const aggregateUserArgsTSFile = await readGeneratedFile(
        "/resolvers/crud/User/args/AggregateUserArgs.ts",
      );
      const findManyUserArgsTSFile = await readGeneratedFile(
        "/resolvers/crud/User/args/FindManyUserArgs.ts",
      );
      const userDistinctFieldEnumTSFile = await readGeneratedFile(
        "/enums/UserDistinctFieldEnum.ts",
      );
      const enumsIndexTSFile = await readGeneratedFile("/enums/index.ts");

      expect(aggregateUserArgsTSFile).toMatchSnapshot("AggregateUserArgs");
      expect(findManyUserArgsTSFile).toMatchSnapshot("FindManyUserArgs");
      expect(userDistinctFieldEnumTSFile).toMatchSnapshot(
        "UserDistinctFieldEnum",
      );
      expect(enumsIndexTSFile).toMatchSnapshot("enums index");
    });
  });

  describe("when `connectOrCreate` is enabled", () => {
    it("should properly generate input type classes for connectOrCreate", async () => {
      const schema = /* prisma */ `
        model User {
          id          Int     @id @default(autoincrement())
          name        String
          postsField  Post[]
        }
        model Post {
          id        Int     @id @default(autoincrement())
          title     String
          authorId  Int
          author    User    @relation(fields: [authorId], references: [id])
        }
      `;

      await generateCodeFromSchema(schema, {
        outputDirPath,
        enabledPreviewFeatures: ["connectOrCreate"],
      });
      const userUpdateOneRequiredWithoutPostsFieldInputTSFile = await readGeneratedFile(
        "/resolvers/inputs/UserUpdateOneRequiredWithoutPostsFieldInput.ts",
      );
      const userCreateOrConnectWithoutpostInputTSFile = await readGeneratedFile(
        "/resolvers/inputs/UserCreateOrConnectWithoutPostInput.ts",
      );
      const userCreateOneWithoutPostsFieldInputTSFile = await readGeneratedFile(
        "/resolvers/inputs/UserCreateOneWithoutPostsFieldInput.ts",
      );
      const inputsIndexTSFile = await readGeneratedFile(
        "/resolvers/inputs/index.ts",
      );

      expect(userUpdateOneRequiredWithoutPostsFieldInputTSFile).toMatchSnapshot(
        "UserUpdateOneRequiredWithoutPostsFieldInput",
      );
      expect(userCreateOrConnectWithoutpostInputTSFile).toMatchSnapshot(
        "UserCreateOrConnectWithoutPostInput",
      );
      expect(userCreateOneWithoutPostsFieldInputTSFile).toMatchSnapshot(
        "UserCreateOneWithoutPostsFieldInput",
      );
      expect(inputsIndexTSFile).toMatchSnapshot("inputs index");
    });
  });
});
