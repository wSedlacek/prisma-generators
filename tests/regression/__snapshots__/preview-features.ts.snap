// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`preview features when \`connectOrCreate\` is enabled should properly generate input type classes for connectOrCreate: UserCreateOneWithoutPostsFieldInput 1`] = `
"import * as TypeGraphQL from \\"type-graphql\\";
import GraphQLJSON from \\"graphql-type-json\\";
import { JsonValue, InputJsonValue } from \\"../../../../helpers/prisma-client-mock\\";
import { UserCreateOrConnectWithoutPostInput } from \\"../inputs/UserCreateOrConnectWithoutPostInput\\";
import { UserCreateWithoutPostsFieldInput } from \\"../inputs/UserCreateWithoutPostsFieldInput\\";
import { UserWhereUniqueInput } from \\"../inputs/UserWhereUniqueInput\\";

@TypeGraphQL.InputType({
  isAbstract: true,
  description: undefined,
})
export class UserCreateOneWithoutPostsFieldInput {
  @TypeGraphQL.Field(_type => UserCreateWithoutPostsFieldInput, {
    nullable: true,
    description: undefined
  })
  create?: UserCreateWithoutPostsFieldInput | undefined;

  @TypeGraphQL.Field(_type => UserWhereUniqueInput, {
    nullable: true,
    description: undefined
  })
  connect?: UserWhereUniqueInput | undefined;

  @TypeGraphQL.Field(_type => UserCreateOrConnectWithoutPostInput, {
    nullable: true,
    description: undefined
  })
  connectOrCreate?: UserCreateOrConnectWithoutPostInput | undefined;
}
"
`;

exports[`preview features when \`connectOrCreate\` is enabled should properly generate input type classes for connectOrCreate: UserCreateOrConnectWithoutPostInput 1`] = `
"import * as TypeGraphQL from \\"type-graphql\\";
import GraphQLJSON from \\"graphql-type-json\\";
import { JsonValue, InputJsonValue } from \\"../../../../helpers/prisma-client-mock\\";
import { UserCreateWithoutPostsFieldInput } from \\"../inputs/UserCreateWithoutPostsFieldInput\\";
import { UserWhereUniqueInput } from \\"../inputs/UserWhereUniqueInput\\";

@TypeGraphQL.InputType({
  isAbstract: true,
  description: undefined,
})
export class UserCreateOrConnectWithoutPostInput {
  @TypeGraphQL.Field(_type => UserWhereUniqueInput, {
    nullable: false,
    description: undefined
  })
  where!: UserWhereUniqueInput;

  @TypeGraphQL.Field(_type => UserCreateWithoutPostsFieldInput, {
    nullable: false,
    description: undefined
  })
  create!: UserCreateWithoutPostsFieldInput;
}
"
`;

exports[`preview features when \`connectOrCreate\` is enabled should properly generate input type classes for connectOrCreate: UserUpdateOneRequiredWithoutPostsFieldInput 1`] = `
"import * as TypeGraphQL from \\"type-graphql\\";
import GraphQLJSON from \\"graphql-type-json\\";
import { JsonValue, InputJsonValue } from \\"../../../../helpers/prisma-client-mock\\";
import { UserCreateOrConnectWithoutPostInput } from \\"../inputs/UserCreateOrConnectWithoutPostInput\\";
import { UserCreateWithoutPostsFieldInput } from \\"../inputs/UserCreateWithoutPostsFieldInput\\";
import { UserUpdateWithoutPostsFieldDataInput } from \\"../inputs/UserUpdateWithoutPostsFieldDataInput\\";
import { UserUpsertWithoutPostsFieldInput } from \\"../inputs/UserUpsertWithoutPostsFieldInput\\";
import { UserWhereUniqueInput } from \\"../inputs/UserWhereUniqueInput\\";

@TypeGraphQL.InputType({
  isAbstract: true,
  description: undefined,
})
export class UserUpdateOneRequiredWithoutPostsFieldInput {
  @TypeGraphQL.Field(_type => UserCreateWithoutPostsFieldInput, {
    nullable: true,
    description: undefined
  })
  create?: UserCreateWithoutPostsFieldInput | undefined;

  @TypeGraphQL.Field(_type => UserWhereUniqueInput, {
    nullable: true,
    description: undefined
  })
  connect?: UserWhereUniqueInput | undefined;

  @TypeGraphQL.Field(_type => UserUpdateWithoutPostsFieldDataInput, {
    nullable: true,
    description: undefined
  })
  update?: UserUpdateWithoutPostsFieldDataInput | undefined;

  @TypeGraphQL.Field(_type => UserUpsertWithoutPostsFieldInput, {
    nullable: true,
    description: undefined
  })
  upsert?: UserUpsertWithoutPostsFieldInput | undefined;

  @TypeGraphQL.Field(_type => UserCreateOrConnectWithoutPostInput, {
    nullable: true,
    description: undefined
  })
  connectOrCreate?: UserCreateOrConnectWithoutPostInput | undefined;
}
"
`;

exports[`preview features when \`connectOrCreate\` is enabled should properly generate input type classes for connectOrCreate: inputs index 1`] = `
"export { IntFilter } from \\"./IntFilter\\";
export { PostCreateInput } from \\"./PostCreateInput\\";
export { PostCreateManyWithoutAuthorInput } from \\"./PostCreateManyWithoutAuthorInput\\";
export { PostCreateOrConnectWithoutUserInput } from \\"./PostCreateOrConnectWithoutUserInput\\";
export { PostCreateWithoutAuthorInput } from \\"./PostCreateWithoutAuthorInput\\";
export { PostFilter } from \\"./PostFilter\\";
export { PostOrderByInput } from \\"./PostOrderByInput\\";
export { PostScalarWhereInput } from \\"./PostScalarWhereInput\\";
export { PostUpdateInput } from \\"./PostUpdateInput\\";
export { PostUpdateManyDataInput } from \\"./PostUpdateManyDataInput\\";
export { PostUpdateManyMutationInput } from \\"./PostUpdateManyMutationInput\\";
export { PostUpdateManyWithWhereNestedInput } from \\"./PostUpdateManyWithWhereNestedInput\\";
export { PostUpdateManyWithoutAuthorInput } from \\"./PostUpdateManyWithoutAuthorInput\\";
export { PostUpdateWithWhereUniqueWithoutAuthorInput } from \\"./PostUpdateWithWhereUniqueWithoutAuthorInput\\";
export { PostUpdateWithoutAuthorDataInput } from \\"./PostUpdateWithoutAuthorDataInput\\";
export { PostUpsertWithWhereUniqueWithoutAuthorInput } from \\"./PostUpsertWithWhereUniqueWithoutAuthorInput\\";
export { PostWhereInput } from \\"./PostWhereInput\\";
export { PostWhereUniqueInput } from \\"./PostWhereUniqueInput\\";
export { StringFilter } from \\"./StringFilter\\";
export { UserCreateInput } from \\"./UserCreateInput\\";
export { UserCreateOneWithoutPostsFieldInput } from \\"./UserCreateOneWithoutPostsFieldInput\\";
export { UserCreateOrConnectWithoutPostInput } from \\"./UserCreateOrConnectWithoutPostInput\\";
export { UserCreateWithoutPostsFieldInput } from \\"./UserCreateWithoutPostsFieldInput\\";
export { UserOrderByInput } from \\"./UserOrderByInput\\";
export { UserUpdateInput } from \\"./UserUpdateInput\\";
export { UserUpdateManyMutationInput } from \\"./UserUpdateManyMutationInput\\";
export { UserUpdateOneRequiredWithoutPostsFieldInput } from \\"./UserUpdateOneRequiredWithoutPostsFieldInput\\";
export { UserUpdateWithoutPostsFieldDataInput } from \\"./UserUpdateWithoutPostsFieldDataInput\\";
export { UserUpsertWithoutPostsFieldInput } from \\"./UserUpsertWithoutPostsFieldInput\\";
export { UserWhereInput } from \\"./UserWhereInput\\";
export { UserWhereUniqueInput } from \\"./UserWhereUniqueInput\\";
"
`;

exports[`preview features when \`distinct\` is enabled should properly generate args classes and enums for model actions resolvers: AggregateUserArgs 1`] = `
"import * as TypeGraphQL from \\"type-graphql\\";
import GraphQLJSON from \\"graphql-type-json\\";
import { UserOrderByInput } from \\"../../../inputs/UserOrderByInput\\";
import { UserWhereInput } from \\"../../../inputs/UserWhereInput\\";
import { UserWhereUniqueInput } from \\"../../../inputs/UserWhereUniqueInput\\";
import { UserDistinctFieldEnum } from \\"../../../../enums/UserDistinctFieldEnum\\";

@TypeGraphQL.ArgsType()
export class AggregateUserArgs {
  @TypeGraphQL.Field(_type => UserWhereInput, { nullable: true })
  where?: UserWhereInput | undefined;

  @TypeGraphQL.Field(_type => UserOrderByInput, { nullable: true })
  orderBy?: UserOrderByInput | undefined;

  @TypeGraphQL.Field(_type => UserWhereUniqueInput, { nullable: true })
  cursor?: UserWhereUniqueInput | undefined;

  @TypeGraphQL.Field(_type => TypeGraphQL.Int, { nullable: true })
  take?: number | undefined;

  @TypeGraphQL.Field(_type => TypeGraphQL.Int, { nullable: true })
  skip?: number | undefined;

  @TypeGraphQL.Field(_type => [UserDistinctFieldEnum], { nullable: true })
  distinct?: Array<keyof typeof UserDistinctFieldEnum> | undefined;
}
"
`;

exports[`preview features when \`distinct\` is enabled should properly generate args classes and enums for model actions resolvers: FindManyUserArgs 1`] = `
"import * as TypeGraphQL from \\"type-graphql\\";
import GraphQLJSON from \\"graphql-type-json\\";
import { UserOrderByInput } from \\"../../../inputs/UserOrderByInput\\";
import { UserWhereInput } from \\"../../../inputs/UserWhereInput\\";
import { UserWhereUniqueInput } from \\"../../../inputs/UserWhereUniqueInput\\";
import { UserDistinctFieldEnum } from \\"../../../../enums/UserDistinctFieldEnum\\";

@TypeGraphQL.ArgsType()
export class FindManyUserArgs {
  @TypeGraphQL.Field(_type => UserWhereInput, { nullable: true })
  where?: UserWhereInput | undefined;

  @TypeGraphQL.Field(_type => UserOrderByInput, { nullable: true })
  orderBy?: UserOrderByInput | undefined;

  @TypeGraphQL.Field(_type => UserWhereUniqueInput, { nullable: true })
  cursor?: UserWhereUniqueInput | undefined;

  @TypeGraphQL.Field(_type => TypeGraphQL.Int, { nullable: true })
  take?: number | undefined;

  @TypeGraphQL.Field(_type => TypeGraphQL.Int, { nullable: true })
  skip?: number | undefined;

  @TypeGraphQL.Field(_type => [UserDistinctFieldEnum], { nullable: true })
  distinct?: Array<keyof typeof UserDistinctFieldEnum> | undefined;
}
"
`;

exports[`preview features when \`distinct\` is enabled should properly generate args classes and enums for model actions resolvers: UserDistinctFieldEnum 1`] = `
"import * as TypeGraphQL from \\"type-graphql\\";

export enum UserDistinctFieldEnum {
  intIdField = \\"intIdField\\",
  uniqueStringField = \\"uniqueStringField\\",
  optionalFloatField = \\"optionalFloatField\\",
  dateField = \\"dateField\\"
}
TypeGraphQL.registerEnumType(UserDistinctFieldEnum, {
  name: \\"UserDistinctFieldEnum\\",
  description: undefined,
});
"
`;

exports[`preview features when \`distinct\` is enabled should properly generate args classes and enums for model actions resolvers: enums index 1`] = `
"export { OrderByArg } from \\"./OrderByArg\\";
export { UserDistinctFieldEnum } from \\"./UserDistinctFieldEnum\\";
"
`;
